ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp

```diff
@@ -0,0 +1,745 @@
+        // TODO: enable this assert
+        // TT_ASSERT(bias_tensor.value().get_shape() == bias_tensor.value().get_legacy_shape());
+    }
+}
+
+ttnn::operations::matmul::MatmulProgramConfig determine_matmul_op_config_from_conv_op_config(
+    OptimizedConvParallelizationConfig conv_parallelization_config,
+    OptimizedConvBlockConfig conv_blocking_config,
+    bool height_sharded,
+    string activation,
+    bool transpose_mcast,
+    uint32_t grid_size_along_c) {
+    if (height_sharded) {
+        ttnn::operations::matmul::MatmulMultiCoreReuseMultiCast1DProgramConfig matmul_config = {
+            .compute_with_storage_grid_size = conv_parallelization_config.grid_size,
+            .in0_block_w = conv_blocking_config.act_block_w_ntiles,
+            .out_subblock_h = conv_blocking_config.out_subblock_h_ntiles,
+            .out_subblock_w = conv_blocking_config.out_subblock_w_ntiles,
+            .per_core_M = conv_parallelization_config.per_core_out_matrix_height_ntiles,
+            .per_core_N = conv_parallelization_config.per_core_out_matrix_width_ntiles,
+            .fuse_batch = true,
+            .mcast_in0 = false};
+        if (activation != "") {
+            matmul_config.fused_activation = ttnn::operations::unary::utils::string_to_unary_with_param(activation);
+        }
+        return matmul_config;
+    } else {
+        TT_ASSERT(conv_blocking_config.act_block_w_ntiles % grid_size_along_c == 0);
+        ttnn::operations::matmul::MatmulMultiCoreReuseMultiCastProgramConfig matmul_config = {
+            .compute_with_storage_grid_size = conv_parallelization_config.grid_size,
+            .in0_block_w = conv_blocking_config.act_block_w_ntiles / grid_size_along_c,
+            .out_subblock_h = conv_blocking_config.out_subblock_h_ntiles,
+            .out_subblock_w = conv_blocking_config.out_subblock_w_ntiles,
+            .per_core_M = conv_parallelization_config.per_core_out_matrix_height_ntiles,
+            .per_core_N = conv_parallelization_config.per_core_out_matrix_width_ntiles,
+            .transpose_mcast = transpose_mcast};
+        if (activation != "") {
+            matmul_config.fused_activation = ttnn::operations::unary::utils::string_to_unary_with_param(activation);
+        }
+        return matmul_config;
+    }
+}
+
+void adjust_conv_op_config_for_auto_shard_if_necessary(
+    bool is_mm_conv,
+    uint32_t batch_size,
+    uint32_t in_channels,
+    uint32_t out_channels,
+    uint32_t output_height,
+    uint32_t output_width,
+    uint32_t weights_width,
+    uint32_t input_width,
+    const CoreCoord& compute_grid_size,
+    Conv2dConfig& conv_config,
+    Layout input_tensor_layout,
+    std::optional<const MemoryConfig> input_memory_config) {
+
+    // If the input tensor is already sharded, or the conv_config has a specified shard layout, we don't need to do anything.
+    if ((input_memory_config.has_value() && input_memory_config.value().is_sharded()) || conv_config.shard_layout.has_value()) {
```

Comment:
this is not the same condition as we had before.
Previously if tensor wasn't on device "is_sharded()" would return false, and we would go into auto shard code path.
With new code, if tensor is host, nullptr will be passed in here and auto shard won't be performed, right?